### ISSUE TITLE
slow WARP evaluation

### ISSUE THREAD

**Issue by christophebjjani** [lyst/lightfm#63]
When evaluating implicit feedbacks (using average mean precision@k for example or even AUC), one needs to predict all users against all items. That is done by looping over the interactions matrix row by row and invoking model.predict()
That gets prohibitively time expensive if faced with a 300K x 20MM matrix. (I estimated ~5days on a 8core server using a simple MF model w/ no content features)
Cython might be able to cut this time by 30%. Sharding the user base and distributing on different servers would reduce that time by another let's say ratio of 10. We would still be looking at 9hours.

What is the way around this? Increasing memory to be able to pass a dense interaction matrix with all ones to model.predict in one or two calls?

**Comment by maciejkula** [comment]
I've opened https://github.com/lyst/lightfm/pull/64 which adds a bunch of optimized evaluation functions, please have a look and see if they help.

In general, evaluating learning-to-rank metrics is expensive because you have to compute the full set of recommendations for every user. If you still find it prohibitively slow I would advise choosing a smaller test set by randomly sampling from your users.

@seanlaw you might find this useful too.

**Comment by seanlaw** [comment]
@maciejkula I was already subscribed for notifications to this issue :)

**Comment by christophebjjani** [comment]
Great, thanks!
How are the hyperparameters tuned (step size, regularization)?
I am thinking of using the same rank estimation approach as in WARP when computing the loss on the validation set to let it scale (specially when doing grid search + CV)

**Comment by maciejkula** [comment]
I'd just use the ranking metric score on the validation set. Especially for WARP watching validation loss (as in neural networks) is not the best approach because the algorithms selects more difficult examples the better it gets, so your raw loss values would be all over the place.

I'd also suggest random search or hyperopt for hyperparams search, but that's up to you :)
